File: go-server/utils/gemini_helper.go
package utils

import (
	"context"
	"encoding/json"
	"fmt"
	"go-server/models"
	"os"
	"regexp"
	"strings"

	"github.com/sirupsen/logrus"
	"google.golang.org/genai"
)

type GeminiHelper struct {
	client   *genai.Client
	apiKey   string
	modelName string
}

// NewGeminiHelper creates a new Gemini helper instance
func NewGeminiHelper() *GeminiHelper {
	apiKey := os.Getenv("API_KEY")
	if apiKey == "" {
		logrus.Warning("No API_KEY found in environment variables. Gemini functionality will not work.")
		return &GeminiHelper{apiKey: ""}
	}

	return &GeminiHelper{
		apiKey:    apiKey,
		modelName: "gemini-2.5-flash", // Default model
	}
}

// InitClient initializes the Gemini client
func (gh *GeminiHelper) InitClient(ctx context.Context) error {
	if gh.apiKey == "" {
		return fmt.Errorf("no API key available for Gemini")
	}

	client, err := genai.NewClient(ctx, nil)
	if err != nil {
		return fmt.Errorf("failed to create Gemini client: %w", err)
	}

	gh.client = client
	return nil
}

// GenerateContent generates content using Gemini API
func (gh *GeminiHelper) GenerateContent(ctx context.Context, prompt string) (string, error) {
	if err := gh.InitClient(ctx); err != nil {
		return "", err
	}
	defer gh.client.Close()

	result, err := gh.client.Models.GenerateContent(
		ctx,
		gh.modelName,
		genai.Text(prompt),
		nil,
	)
	if err != nil {
		logrus.Errorf("Error generating content: %v", err)
		return "", err
	}

	return result.Text(), nil
}

// GenerateContentWithThinking generates content with thinking disabled
func (gh *GeminiHelper) GenerateContentWithThinking(ctx context.Context, prompt string, disableThinking bool) (string, error) {
	if err := gh.InitClient(ctx); err != nil {
		return "", err
	}
	defer gh.client.Close()

	var config *genai.GenerateContentConfig
	if disableThinking {
		config = &genai.GenerateContentConfig{
			ThinkingConfig: &genai.ThinkingConfig{
				ThinkingBudget: int32(0), // Disables thinking
			},
		}
	}

	result, err := gh.client.Models.GenerateContent(
		ctx,
		gh.modelName,
		genai.Text(prompt),
		config,
	)
	if err != nil {
		logrus.Errorf("Error generating content with thinking: %v", err)
		return "", err
	}

	return result.Text(), nil
}

// GenerateContentWithSystemInstruction generates content with system instructions
func (gh *GeminiHelper) GenerateContentWithSystemInstruction(ctx context.Context, prompt, systemInstruction string) (string, error) {
	if err := gh.InitClient(ctx); err != nil {
		return "", err
	}
	defer gh.client.Close()

	config := &genai.GenerateContentConfig{
		SystemInstruction: genai.NewContentFromText(systemInstruction, genai.RoleUser),
	}

	result, err := gh.client.Models.GenerateContent(
		ctx,
		gh.modelName,
		genai.Text(prompt),
		config,
	)
	if err != nil {
		logrus.Errorf("Error generating content with system instruction: %v", err)
		return "", err
	}

	return result.Text(), nil
}

// GenerateStructuredOutput generates JSON output using response schema
func (gh *GeminiHelper) GenerateStructuredOutput(ctx context.Context, prompt string, schema *genai.Schema) (string, error) {
	if err := gh.InitClient(ctx); err != nil {
		return "", err
	}
	defer gh.client.Close()

	config := &genai.GenerateContentConfig{
		ResponseMIMEType: "application/json",
		ResponseSchema:   schema,
	}

	result, err := gh.client.Models.GenerateContent(
		ctx,
		gh.modelName,
		genai.Text(prompt),
		config,
	)
	if err != nil {
		logrus.Errorf("Error generating structured output: %v", err)
		return "", err
	}

	return result.Text(), nil
}

// GenerateSubjects generates subjects for a course using structured output
func (gh *GeminiHelper) GenerateSubjects(ctx context.Context, courseName, courseDescription string) ([]string, error) {
	prompt := fmt.Sprintf(`
Generate a comprehensive list of subjects for the course "%s" with description: "%s".
Provide 5-8 core subjects that would be essential for this course.
Return only the subject names as a JSON array.
`, courseName, courseDescription)

	schema := &genai.Schema{
		Type: genai.TypeArray,
		Items: &genai.Schema{
			Type: genai.TypeString,
		},
	}

	result, err := gh.GenerateStructuredOutput(ctx, prompt, schema)
	if err != nil {
		return nil, err
	}

	var subjects []string
	if err := json.Unmarshal([]byte(result), &subjects); err != nil {
		logrus.Errorf("Error unmarshaling subjects: %v", err)
		return nil, err
	}

	return subjects, nil
}

// GenerateChapters generates chapters for a subject using structured output
func (gh *GeminiHelper) GenerateChapters(ctx context.Context, subjectName, courseName string) ([]string, error) {
	prompt := fmt.Sprintf(`
Generate a comprehensive list of chapters for the subject "%s" in the course "%s".
Provide 6-10 logical chapters that would cover this subject thoroughly.
Return only the chapter names as a JSON array.
`, subjectName, courseName)

	schema := &genai.Schema{
		Type: genai.TypeArray,
		Items: &genai.Schema{
			Type: genai.TypeString,
		},
	}

	result, err := gh.GenerateStructuredOutput(ctx, prompt, schema)
	if err != nil {
		return nil, err
	}

	var chapters []string
	if err := json.Unmarshal([]byte(result), &chapters); err != nil {
		logrus.Errorf("Error unmarshaling chapters: %v", err)
		return nil, err
	}

	return chapters, nil
}

// GenerateTopics generates topics for a chapter using structured output
func (gh *GeminiHelper) GenerateTopics(ctx context.Context, chapterName, subjectName, courseName string) ([]string, error) {
	prompt := fmt.Sprintf(`
Generate a comprehensive list of topics for the chapter "%s" in subject "%s" for the course "%s".
Provide 5-8 specific topics that would be covered in this chapter.
Return only the topic names as a JSON array.
`, chapterName, subjectName, courseName)

	schema := &genai.Schema{
		Type: genai.TypeArray,
		Items: &genai.Schema{
			Type: genai.TypeString,
		},
	}

	result, err := gh.GenerateStructuredOutput(ctx, prompt, schema)
	if err != nil {
		return nil, err
	}

	var topics []string
	if err := json.Unmarshal([]byte(result), &topics); err != nil {
		logrus.Errorf("Error unmarshaling topics: %v", err)
		return nil, err
	}

	return topics, nil
}

// GenerateTopicContent generates detailed content for a topic
func (gh *GeminiHelper) GenerateTopicContent(ctx context.Context, topicName, chapterName, subjectName, courseName string) (string, error) {
	prompt := fmt.Sprintf(`
Generate detailed, in-depth content as well as a tutorial for the topic "%s", under the chapter "%s",
under the subject "%s" for the course "%s".

The content should adhere to the following detailed approach:

1. **Introduction**: Start with a clear and engaging introduction that outlines what will be covered in this topic.

2. **Core Concepts**: Break down the topic into fundamental concepts, explaining each with clear definitions and examples.

3. **Detailed Explanation**: Provide comprehensive explanations with:
   - Real-world examples and applications
   - Step-by-step processes where applicable
   - Common misconceptions and how to avoid them
   - Best practices and industry standards

4. **Practical Examples**: Include multiple practical examples that demonstrate the concepts in action.

5. **Interactive Elements**: Where appropriate, include:
   - Code examples (if technical)
   - Diagrams descriptions
   - Case studies
   - Exercise suggestions

6. **Summary**: Conclude with a concise summary highlighting the key takeaways.

7. **Further Reading**: Suggest additional resources for deeper learning.

Format the content using markdown with proper headings, bullet points, code blocks (where applicable), and emphasis.
Make it comprehensive, educational, and engaging for learners.
`, topicName, chapterName, subjectName, courseName)

	content, err := gh.GenerateContent(ctx, prompt)
	if err != nil {
		return "", err
	}

	// Process the content to handle special formatting
	processedContent := gh.ProcessMermaidContent(content)
	processedContent = gh.ExtractMarkdown(processedContent)

	return processedContent, nil
}

// ProcessMermaidContent processes mermaid diagrams in the content
func (gh *GeminiHelper) ProcessMermaidContent(content string) string {
	// Find and replace mermaid code blocks
	mermaidRegex := regexp.MustCompile(`(?s)` + "`" + `{3}mermaid(.*?)` + "`" + `{3}`)
	
	return mermaidRegex.ReplaceAllStringFunc(content, func(match string) string {
		// Extract the content between ```mermaid and ```
		lines := strings.Split(match, "\n")
		if len(lines) < 2 {
			return match
		}
		
		// Remove the first line (```mermaid) and last line (```)
		mermaidContent := strings.Join(lines[1:len(lines)-1], "\n")
		return fmt.Sprintf(`<pre class="mermaid">%s</pre>`, mermaidContent)
	})
}

// ExtractMarkdown removes markdown wrapper if present
func (gh *GeminiHelper) ExtractMarkdown(content string) string {
	// Remove the first occurrence of ```markdown and last occurrence of ```
	if strings.Contains(content, "```markdown") {
		start := strings.Index(content, "```markdown")
		end := strings.LastIndex(content, "```")
		if start != -1 && end != -1 && end > start {
			return strings.TrimSpace(content[start+12 : end])
		}
	}
	return content
}

// ExtractSQLQuery extracts SQL query from response text
func (gh *GeminiHelper) ExtractSQLQuery(responseText string) (string, error) {
	// Find the start and end of the SQL block
	sqlStart := strings.Index(responseText, "```sql")
	if sqlStart == -1 {
		return "", fmt.Errorf("SQL block not found in the response")
	}
	sqlStart += 6 // Skip "```sql"

	sqlEnd := strings.Index(responseText[sqlStart:], "```")
	if sqlEnd == -1 {
		return "", fmt.Errorf("SQL block end not found in the response")
	}

	// Extract and clean the SQL query
	sqlQuery := strings.TrimSpace(responseText[sqlStart : sqlStart+sqlEnd])

	// Remove leading 'l' if present (cleanup from the Python implementation)
	if strings.HasPrefix(sqlQuery, "l") {
		sqlQuery = strings.TrimSpace(sqlQuery[1:])
	}

	return sqlQuery, nil
}

// GenerateImageWithText generates image based on descriptive prompt (for Gemini 2.0 models)
func (gh *GeminiHelper) GenerateImageWithText(ctx context.Context, prompt string) ([]byte, string, error) {
	if err := gh.InitClient(ctx); err != nil {
		return nil, "", err
	}
	defer gh.client.Close()

	config := &genai.GenerateContentConfig{
		ResponseModalities: []string{"TEXT", "IMAGE"},
	}

	result, err := gh.client.Models.GenerateContent(
		ctx,
		"gemini-2.0-flash-preview-image-generation", // Image generation model
		genai.Text(prompt),
		config,
	)
	if err != nil {
		logrus.Errorf("Error generating image with text: %v", err)
		return nil, "", err
	}

	var imageBytes []byte
	var textResponse string

	for _, part := range result.Candidates[0].Content.Parts {
		if part.Text != "" {
			textResponse = part.Text
		} else if part.InlineData != nil {
			imageBytes = part.InlineData.Data
		}
	}

	return imageBytes, textResponse, nil
}

File: python-server/utils/gemini_helper.py
# utils/gemini_helper.py
import os
import logging
import json
from google import genai
from google.genai import types
from config import Config
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

logger = logging.getLogger(__name__)

def extract_sql_query(response_text):
    try:
        # Find the start and end of the SQL block
        sql_block_start = response_text.find("```sql") + 6  # Start after "```sql"
        sql_block_end = response_text.find("```", sql_block_start)

        if sql_block_start == -1 or sql_block_end == -1:
            raise ValueError("SQL block not found in the response")

        # Extract and clean the SQL query
        sql_query = response_text[sql_block_start:sql_block_end].strip()

        # Ensure the SQL query is valid (remove unwanted characters)
        if sql_query.startswith("l"):
            sql_query = sql_query[1:].strip()  # Remove the leading 'l' if present

        return sql_query
    except Exception as e:
        raise ValueError(f"Error extracting SQL query: {str(e)}")


def mermaid_content(content):
    """Convert mermaid code blocks to HTML pre tags"""
    start = content.find('```mermaid')
    end = content.find('```', start + 1)
    while start != -1 and end != -1:
        mermaid_code = content[start:end + 3]
        mermaid_code = mermaid_code.replace('```mermaid', '<pre class="mermaid">').replace('```', '</pre>')
        content = content[:start] + mermaid_code + content[end + 3:]
        start = content.find('```mermaid', end + 1)
        end = content.find('```', start + 1)
    return content


def chart_content(content):
    """Convert chart-json code blocks to HTML divs with data attributes"""
    import re
    import html
    
    # Pattern to match ```chart-json ... ```
    pattern = r'```chart-json\s*\n(.*?)```'
    
    def replace_chart(match):
        chart_json = match.group(1).strip()
        # Generate unique ID for each chart
        import time
        chart_id = f"chart-{int(time.time() * 1000)}-{hash(chart_json) % 10000}"
        # Escape the JSON for HTML attribute
        escaped_json = html.escape(chart_json)
        return f'<div class="chart-container" data-chart-id="{chart_id}" data-chart-config="{escaped_json}"></div>'
    
    # Replace all chart-json blocks
    content = re.sub(pattern, replace_chart, content, flags=re.DOTALL)
    return content


# remove the first occurence of  ```markdown  and last occurence of ```
def extract_markdown(content):
    start = content.find('```markdown')
    end = content.rfind('```')
    if start != -1 and end != -1:
        content = content[start + 12:end].strip()
    return content





class GeminiHelper:
    def __init__(self, api_key=None):
        # Always use API key from environment variables
        self.api_key = os.environ.get('API_KEY')
        
        if not self.api_key:
            logger.warning("No API_KEY found in environment variables. Gemini functionality will not work.")
        
        # Initialize the client only when needed, with API key
        self._client = None
        self.safety_settings = [
            types.SafetySetting(
                category="HARM_CATEGORY_SEXUALLY_EXPLICIT",
                threshold="BLOCK_NONE"
            ),
            types.SafetySetting(
                category="HARM_CATEGORY_HATE_SPEECH",
                threshold="BLOCK_NONE"
            ),
            types.SafetySetting(
                category="HARM_CATEGORY_HARASSMENT",
                threshold="HIGH"
            ),
            types.SafetySetting(
                category="HARM_CATEGORY_DANGEROUS_CONTENT",
                threshold="HIGH"
            )
        ]
        self.generation_config = {
            "max_output_tokens": 8192
        }
        self.model_name = "gemini-2.5-flash-lite"  # Default model

    @property
    def client(self):
        # Only initialize the client when needed, not in constructor
        if self._client is None:
            if not self.api_key:
                raise ValueError("No API key available for Gemini. Set API_KEY in environment variables.")
            self._client = genai.Client(api_key=self.api_key)
        return self._client

    def generate_content(self, prompt, response_schema=None):
        try:
            logger.debug(f"Generating content with schema: {response_schema}")
            if not self.api_key:
                raise ValueError("No API key available for Gemini. Set API_KEY in environment variables.")
                
            if response_schema:
                config = types.GenerateContentConfig(
                    response_mime_type="application/json",
                    response_schema=response_schema,
                    max_output_tokens=self.generation_config["max_output_tokens"],
                    # safety_settings=self.safety_settings
                )
                logger.debug(f"Using generation config: {config}")
                
                response = self.client.models.generate_content(
                    model=self.model_name,
                    contents=prompt,
                    config=config
                )
                logger.debug(f"Raw response from Gemini: {response.text}")
                return json.loads(response.text)
            else:
                response = self.client.models.generate_content(
                    model=self.model_name, 
                    contents=prompt,
                    config=types.GenerateContentConfig(
                        # safety_settings=self.safety_settings
                    )
                )
                return response.text
        except Exception as e:
            logger.error(f"Error in generate_content: {str(e)}", exc_info=True)
            raise
        finally:
            # Remove reference to client to help with garbage collection
            # of objects containing the API key
            self._client = None

